hicommonwealth/commonwealth
# POC of new topic gating flow

## Link to Issue
Closes: #4460 

## Description of Changes
- Deprecates dependence on `ITokenAdapters`balance checks
- Deprecates use of all functions within `gateTopic.ts` 
- adds lightweight `gateStrategies` data structure to `/bulkOffchain` 
    - Loads gateStrategies(topic thresholds, etc) into client memory 1 time upon chain initialization
    - Helper functions on the chain object for checking gates and thresholds 
- Adds single `getTokenBalance` helper function to client:
    - Only executes when gates exist and are > 0 in a community
    - populates `user.activeAccounts` with all user balances for all addresses on given chain
- New code path and request param(`all: boolean`) added to `/tokenBalance` which will get all balances for the user of an address across addresses
- async token balance request occur when setting active accounts(anytime you enter a community page)     

## Test Plan
- Using an ethereum account with some ERC20 token:
    - Create a community or use existing community as admin
    - set up a token threshold
    - Join community with address holding token
    - ensure create thread threshold passes
    - check comment and confirm balance is correct
    - move topic threshold above the balance of this address
    - ensure all actions are now blocked
- Go to cosmos communities with keplr accounts and ensure balances and gates are correct(ie evmos should be restrictive of almost all topics with a 0 balance)
- attempt to join/go to non threshold communities and confirm no errors
